lib/auth/recordingencryption/recordingencryptionv1/service.go

```diff
@@ -195,9 +218,59 @@ func (s *Service) CompleteUpload(ctx context.Context, req *recordingencryptionv1
 		return nil, trace.Wrap(err)
 	}
 
+	sessionEnd, err := s.findSessionEndEvent(ctx, upload.SessionID)
+	if err != nil || sessionEnd == nil {
+		return &recordingencryptionv1.CompleteUploadResponse{}, nil
+	}
+
+	summarizer := s.sessionSummarizerProvider.SessionSummarizer()
+	switch o := sessionEnd.(type) {
+	case *apievents.SessionEnd:
+		if err := summarizer.SummarizeSSH(ctx, o); err != nil {
+			s.logger.WarnContext(ctx, "failed to summarize upload", "error", err)
+		}
+		metadataSvc := s.recordingMetadataProvider.Service()
+		if !o.EndTime.IsZero() && !o.StartTime.IsZero() {
+			duration := o.EndTime.Sub(o.StartTime)
+			if err := metadataSvc.ProcessSessionRecording(ctx, upload.SessionID, duration); err != nil {
+				s.logger.WarnContext(ctx, "failed to process session recording metadata", "error", err)
+			}
+		}
+	case *apievents.DatabaseSessionEnd:
+		if err := summarizer.SummarizeDatabase(ctx, o); err != nil {
+			s.logger.WarnContext(ctx, "failed to summarize upload", "error", err)
+		}
+	}
+
 	return &recordingencryptionv1.CompleteUploadResponse{}, nil
 }
 
+// findSessionEndEvent streams session events to find the session end event for the given session ID.
+// It returns either a SessionEnd or DatabaseSessionEnd event, or nil if none is found.
+func (s *Service) findSessionEndEvent(ctx context.Context, sessionID session.ID) (apievents.AuditEvent, error) {
```

Comment:
This looks very similar to other summarization logic. Is there any way to avoid the duplication?